### PR TITLE
add spring-controller sub-generator

### PR DESCRIPTION
**Pull request by xetys**
for #6440 

based on service and renamed service subgen to spring-service

I would like to add both, travis tests and integration tests for the controller...

additionally, it might be interesting for the interactive mode to add controller actions like fields in entity generator, where the user can choose the HTTP method and the action's name. WDYT? 

- [x] [Travis tests](https://travis-ci.org/jhipster/generator-jhipster/pull_requests) are green
- [x] Tests are added where necessary
- [ ] Documentation is added/updated where necessary
- [x] Coding Rules & Commit Guidelines as per our [CONTRIBUTING.md document](https://github.com/jhipster/generator-jhipster/blob/master/CONTRIBUTING.md) are followed

### CODE REVIEW

File: cli/commands.js [hunk 1]
```diff
@@ -1,62 +1,66 @@
 module.exports = {
-    app: {
+    'app': {
```
Comment (MathieuAA) [inline cli/commands.js]: Why did you add these? They actually make the code a bit harder to read and are unnecessary (except for `spring-controller` where it's needed).

**Comment by xetys** [comment]
No, there are several other, as 'import-jdl', 'docker-compose' etc. I found
it inconsistent to read so I unified it... I can take it back if it's needed

Am Mo., 2. Okt. 2017, 08:35 schrieb Mathieu ABOU-AICHI <
notifications@github.com>:

**Comment by MathieuAA** [comment]
@xetys that's the trade-off =/ it's generally the linter that warns you (more of an error-warning :)) about it.
It seems, in our case, that the tests fail because of that.

File: cli/commands.js [hunk 2]
```diff
@@ -1,62 +1,66 @@
...
     },
     'rancher-compose': {
         desc: 'Deploy the current application to Rancher'
     },
-    server: {
+    'server': {
         desc: 'Create a new JHipster server-side application'
     },
-    service: {
+    'spring-service': {
```
Comment (deepu105) [inline cli/commands.js]: add `alias : 'service'` here so that old command works as well, else this will be breaking change

**Comment by xetys** [comment]
@deepu105 seriously...ok... I will change that

File: generators/spring-controller/templates/src/main/java/package/web/rest/_Controller.java [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+
+/**
+ * <%= controllerClass %> controller
+ */
+@RestController
+@RequestMapping("/api/<%= apiPrefix %>")
+public class <%= controllerClass %>Controller {
+
+    private final Logger log = LoggerFactory.getLogger(<%= controllerClass %>Controller.class);
+
```
Comment (jdubois) [inline generators/spring-controller/templates/src/main/java/package/web/rest/_Controller.java]: Maybe we should also add a simple "GET" method? Like a "helloworld()" method that returns a String?

File: generators/spring-controller/templates/src/main/java/package/web/rest/_Controller.java [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RestController;
+
+/**
+ * <%= controllerClass %> controller
+ */
+@RestController
+@RequestMapping("/api/<%= apiPrefix %>")
```
Comment (jdubois) [inline generators/spring-controller/templates/src/main/java/package/web/rest/_Controller.java]: Maybe we should also generate a unique URL? Here you would get error when you generate 2 controllers in a row. How about "/api/<%= apiPrefix %>-<%=controllerClass%>"

Comment (jdubois) [reply generators/spring-controller/templates/src/main/java/package/web/rest/_Controller.java]: Maybe use Lodash to use a dasherized version of "controllerClass"

File: generators/spring-controller/templates/src/main/java/package/web/rest/_Controller.java (hunk 3 above)
Comment (xetys) [reply generators/spring-controller/templates/src/main/java/package/web/rest/_Controller.java]: How about a similar process like the field adding in entity sub-generator? And if no method added, then add a hello-world method

File: generators/spring-controller/templates/src/main/java/package/web/rest/_Controller.java (hunk 4 above)
Comment (xetys) [reply generators/spring-controller/templates/src/main/java/package/web/rest/_Controller.java]: apiPrefix = kebabCase(controllerClass)

**Comment by xetys** [comment]
my todos left:

- [ ]: yeoman tests
- [ ]: Spring integration tests
- [ ]: (in discusson) method generator
- [ ]: default action (if no other action exists)

File: generators/spring-controller/templates/src/main/java/package/web/rest/_Controller.java (hunk 3 above)
Comment (deepu105) [reply generators/spring-controller/templates/src/main/java/package/web/rest/_Controller.java]: That would be much better otherwise the class we generate wouldnt have much value over the one you could create in an IDE

Comment (jdubois) [reply generators/spring-controller/templates/src/main/java/package/web/rest/_Controller.java]: yes +1 - but keep the question simple

File: generators/spring-controller/templates/src/main/java/package/web/rest/_Controller.java (hunk 4 above)
Comment (jdubois) [reply generators/spring-controller/templates/src/main/java/package/web/rest/_Controller.java]: the issue with KebabCase is that you would keep upper case letters -> maybe do a lower case after?

Comment (xetys) [reply generators/spring-controller/templates/src/main/java/package/web/rest/_Controller.java]: no, it doesn't, I tried already...

```
/**
 * AnotherTest controller
 */
@RestController
@RequestMapping("/api/another-test")
public class AnotherTestController {
    //...
}
```

Comment (jdubois) [reply generators/spring-controller/templates/src/main/java/package/web/rest/_Controller.java]: I'm quite surprised, good news then, let's do it!

**Comment by xetys** [comment]
so far, here it is

File: generators/spring-controller/index.js [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+
+        this.controllerActions.forEach((action) => {
+            action.actionPath = _.kebabCase(action.actionName);
+            action.actionNameUF = _.upperFirst(action.actionName);
+            this.log(chalk.green(`adding ${action.actionMethod} action '${action.actionName}' for /api/${this.apiPrefix}/${action.actionPath}`));
+        });
+
+        this.template(
+            `${SERVER_MAIN_SRC_DIR}package/web/rest/_Controller.java`,
+            `${SERVER_MAIN_SRC_DIR + this.packageFolder}/web/rest/${this.controllerClass}Controller.java`
```
Comment (deepu105) [inline generators/spring-controller/index.js]: this could be `${SERVER_MAIN_SRC_DIR}${this.packageFolder}/web/rest/${this.controllerClass}Controller.java`
same in other usage below

File: generators/spring-controller/prompts.js [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+    askForControllerActions
+};
+
+function askForControllerActions() {
+    const done = this.async();
+
+    askForControllerAction.call(this, done);
+}
+
+function askForControllerAction(done) {
```
Comment (deepu105) [inline generators/spring-controller/prompts.js]: You can move this method inside the `askForControllerActions` and convert this into a const arrow like below so that you dont have to bind `this`
```
function askForControllerActions() {
  const askForControllerAction = done => {
    ...
    this.prompt(prompts).then((props) => {
      if (props.actionAdd) {
        ...
        askForControllerAction(done);
      }
    });
  }
  const done = this.async();
  askForControllerAction(done);
}
```

File: generators/spring-controller/prompts.js [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+                },
+                {
+                    name: 'PUT',
+                    value: 'Put'
+                },
+                {
+                    name: 'DELETE',
+                    value: 'Delete'
+                }
+
```
Comment (jdubois) [inline generators/spring-controller/prompts.js]: Please remove the empty spaces

File: generators/spring-controller/templates/src/test/java/package/web/rest/_ControllerIntTest.java [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+import <%= mockRequestImports[idx] %>;
+<%_ } _%>
+import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;
+/**
+ * Test class for the <%= controllerClass %> REST controller.
+ *
+ * @see <%= controllerClass %>Controller
+ */
+@RunWith(SpringRunner.class)
+@SpringBootTest(classes = <%= mainClass %>.class)
```
Comment (jdubois) [inline generators/spring-controller/templates/src/test/java/package/web/rest/_ControllerIntTest.java]: Oh you even generate tests on your controller!!! Big +1!

File: generators/spring-controller/prompts.js [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+                type: 'input',
+                name: 'actionName',
+                validate: (input) => {
+                    if (!(/^([a-zA-Z0-9_]*)$/.test(input))) {
+                        return 'Your action name cannot contain special characters';
+                    } else if (input === '') {
+                        return 'Your action name cannot be empty';
+                    } else if (input.charAt(0) === input.charAt(0).toUpperCase()) {
+                        return 'Your action name cannot start with an upper case letter';
+                    } else if (jhiCore.isReservedFieldName(input)) {
```
Comment (deepu105) [inline generators/spring-controller/prompts.js]: @xetys does this matter? I guess reserved words are ok in a method name??

Comment (deepu105) [reply generators/spring-controller/prompts.js]: Ok never mind seems some keywords cant be used anyway

File: generators/spring-controller/prompts.js [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+        this.prompt(prompts).then((props) => {
+            if (props.actionAdd) {
+                const controllerAction = {
+                    actionName: props.actionName,
+                    actionMethod: props.actionMethod
+                };
+
+                this.controllerActions.push(controllerAction);
+
+                askForControllerAction.call(this, done);
```
Comment (deepu105) [inline generators/spring-controller/prompts.js]: you cant do this for arrow methods just call `askForControllerAction(done)`

**Comment by jdubois** [comment]
I'm rushing to do the `4.10.0` release, so sorry I'm going to modify this without much discussion:

- This generates a "FooController" class, but the entity sub-generator would generate a "FooResource" class. This is because that's supposed to be a REST "resource". Anyway, for consistency, I'm going to rename this with the "Resource" prefix.
- I'm also going to add the documentation to the website.
